libmount/src/utils.c

```diff
@@ -1342,6 +1348,207 @@ int mnt_tmptgt_cleanup(int old_ns_fd)
 #endif
 }
 
+#ifdef UL_HAVE_MOUNT_API
+static int write_id_mapping(idmap_type_t idtype, pid_t pid, const char *buf,
+			    size_t buf_size)
+{
+	int fd = -1, ret = -1, setgroups_fd = -1;
+	char path[PATH_MAX];
+
+	if (geteuid() != 0 && idtype == ID_TYPE_GID) {
+		snprintf(path, sizeof(path), "/proc/%d/setgroups", pid);
+
+		setgroups_fd = open(path, O_WRONLY | O_CLOEXEC | O_NOCTTY);
+		if (setgroups_fd < 0 && errno != ENOENT)
+			goto err;
+
+		if (setgroups_fd >= 0) {
+			ret = write(setgroups_fd, "deny\n", strlen("deny\n"));
+			if (ret < 0 || ret != strlen("deny\n"))
+				goto err;
+		}
+	}
+
+	snprintf(path, sizeof(path), "/proc/%d/%cid_map", pid,
+		 idtype == ID_TYPE_UID ? 'u' : 'g');
+
+	fd = open(path, O_WRONLY | O_CLOEXEC | O_NOCTTY);
+	if (fd < 0)
+		goto err;
+
+	ret = write(fd, buf, buf_size);
+	if (ret < 0 || (size_t)ret != buf_size)
+		goto err;
+
+	ret = 0;
+err:
+	if (fd >= 0)
+		close(fd);
+	if (setgroups_fd >= 0)
+		close(setgroups_fd);
+
+	return ret;
+}
+
+static int map_ids(struct list_head *idmap, pid_t pid)
+{
+	int fill, left;
+	char *pos;
+	int ret = 0;
+	char mapbuf[4096] = {};
+	struct list_head *p;
+
+	for (idmap_type_t type = ID_TYPE_UID; type <= ID_TYPE_GID; type++) {
+		bool had_entry = false;
+
+		pos = mapbuf;
+		list_for_each(p, idmap) {
+			struct id_map *map = list_entry(p, struct id_map, map_head);
+
+			/*
+			 * If the map type is ID_TYPE_UIDGID we need to include
+			 * it in both gid- and uidmap.
+			 */
+			if (map->map_type != ID_TYPE_UIDGID && map->map_type != type)
+				continue;
+
+			had_entry = true;
+
+			left = sizeof(mapbuf) - (pos - mapbuf);
+			fill = snprintf(pos, left,
+					"%" PRIu32 "%" PRIu32 "%" PRIu32 "\n",
+					map->nsid, map->hostid, map->range);
+			/*
+			 * The kernel only takes <= 4k for writes to
+			 * /proc/<pid>/{g,u}id_map
+			 */
+			if (fill <= 0)
+				return errno = EINVAL, -1;
+
+			pos += fill;
+		}
+		if (!had_entry)
+			continue;
+
+		ret = write_id_mapping(type, pid, mapbuf, pos - mapbuf);
+		if (ret < 0)
+			return -1;
+
+		memset(mapbuf, 0, sizeof(mapbuf));
+	}
+
+	return 0;
+}
+
+static int wait_for_pid(pid_t pid)
+{
+	int status, ret;
+
+again:
```

Comment:
dunno about util-linux, but in systemd we have the rule that goto may not be used as to jump up, i.e. to build loops. only Ok to jump to some cleanup label...

but @karelzak is the one defining coding style here...

Reply:
I mean, I can just make this a do-while construct.